apps/prairielearn/src/ee/middlewares/checkPlanGrants.ts

```diff
@@ -0,0 +1,14 @@
+import asyncHandler = require('express-async-handler');
+
+import { checkPlanGrants } from '../lib/billing/plan-grants';
+
+export default asyncHandler(async (req, res, next) => {
```

Comment:
We could consider adding this middleware after this in the `authzWorkspace` middleware:

https://github.com/PrairieLearn/PrairieLearn/blob/96e5ad650aec558bdb2475df3b6fbafd909f3b83/apps/prairielearn/src/middlewares/authzWorkspace.js#L28

Reply:
I'm not sure I understand the implication of doing this?

Reply:
Consider the following contrived scenario:

- Student enrolls and pays
- Student student opens a workspace question and saves the URL
- Student requests a refund; we grant it and remove the plan grant
- Student visits the workspace URL

Currently, I _think_ the student will still have access to it, as the new `checkPlanGrants` middleware won't have run.

Alternatively: we can move this check into the `authzCourseOrInstance` middleware so that we get this check automatically every time we authz a course instance.

Reply:
I think I'd prefer to move it into `authzCourseOrInstance` as the single source of checking authz.

Reply:
Ah, ok, I remember now why I didn't put it in `authzCourseOrInstance`: the `/course_instance/:id/upgrade` route specifically needs this middleware to *not* run, otherwise we'll get stuck in a loop of that page redirecting back to itself over and over again.

`authzCourseOrInstance` is basically the first middleware we mount on `/pl/course_instance/:id`, so it would be a pain to reorder things. Plus, thinking out loud, the upgrade page actually *needs* `authzCourseOrInstance` to have run so that we can resolve the effective user information!

I think the best thing to do for now will be to just add this middleware to the workspace auth middleware. I think that's the only place that we serve course instance content to students where the course instance isn't actually in the route.

The alternative is to allowlist the `.../upgrade` route in the middleware... which is maybe not terrible? I'm not sure!

Reply:
Makes sense to me! I'm fine to either add the middleware to the workspace auth middleware (or mount it just before), or to allowlist `upgrade/`. The allowlist feels a bit more hacky to me? I guess I'm a bit worried about having too many special cases hidden inside different bits of authz code. But this is a pretty special case so I'm not too worried here.

Reply:
Yeah, I don't love adding an allowlist there. I think adding this to the workspace auth middleware is probably the best option - that's *already* a special case itself, so it'll require a little more care for any change we make, including this one.

Reply:
Changed in 91c19af. I imported and called the middleware in the `authzWorkspace` middleware, so it'll work for both the "shell" page that loads the iframe and any requests to the underlying container.

Reply:
Looks good!